lib/logstash/config/config_ast.rb

```diff
@@ -309,11 +319,36 @@ def compile
     end
   end
   class Hash < Value
+    def validate!
+      duplicate_values = find_duplicate_keys
+
+      if duplicate_values.size > 0
+        raise ConfigurationError.new(
+          I18n.t("logstash.agent.configuration.invalid_plugin_settings_duplicated_keys",
+                 :keys => duplicate_values.join(', '),
+                 :line => input.line_of(interval.first),
+                 :column => input.column_of(interval.first),
+                 :byte => interval.first + 1,
+                 :after => input[0..interval.first]
+                )
+        )
+      end
+    end
+
+    def find_duplicate_keys
+      values = recursive_select(HashEntry).collect { |hash_entry| hash_entry.name.text_value_utf8 }
+      values.find_all { |v| values.count(v) > 1 }.uniq
+    end
```

Comment:
its more about understanding the `recursive_select(HashEntry)` behaviour. The way it reads is that  it will find all keys recursively and create a flat array of all keys with count and if so my previous example would fail - but it does not, I actually tested it. 

So, bottom line is that it is hard to understand what this does exactly, maybe add a explanation comment on what is happening?